dask/core.py

```diff
@@ -224,29 +230,41 @@ def get_dependencies(dsk, key=None, task=None, as_list=False):
     set(['x'])
     """
     if key is not None:
-        args = [dsk[key]]
+        arg = dsk[key]
     elif task is not None:
-        args = [task]
+        arg = task
     else:
         raise ValueError("Provide either key or task")
-    result = []
-    while args:
-        arg = args.pop()
-        typ = type(arg)
-        if typ is tuple and arg and callable(arg[0]):  # istask(arg):
-            args.extend(arg[1:])
-        elif type(arg) is list:
-            args.extend(arg)
-        else:
-            result.append(arg)
-    if not result:
-        return [] if as_list else set()
-    rv = []
-    for x in result:
-        rv.extend(_deps(dsk, x))
+    rv = _deps(dsk, arg)
     return rv if as_list else set(rv)
 
 
+def get_all_dependencies(dsk, tasks):
+    """ Get the immediate tasks on which the tasks depend.
+
+    Like get_dependencies(), but operating on multiple tasks at once,
+    and therefore significantly faster.
+    """
+    if not isinstance(tasks, list):
+        raise TypeError("Please provide a list of tasks")
+    return set(_deps(dsk, tasks))
+
+
+def get_all_dependencies_per_key(dsk, keys, as_list=False):
+    """ Get the immediate tasks on which the keys depend,
+    as a dictionary.
+
+    Like get_dependencies(), but operating on multiple tasks at once,
+    and therefore significantly faster.
+    """
+    if not isinstance(keys, list):
+        raise TypeError("Please provide a list of keys")
+    if as_list:
+        return {k: _deps(dsk, dsk[k]) for k in keys}
+    else:
+        return {k: set(_deps(dsk, dsk[k])) for k in keys}
```

Comment:
This can also happen later.  Mostly I want to push on the principle of "there is a maintenance cost to having lots of functions that do similar things".  If we can have a few of these that compose well with standard language constructs like dictionary comprehensions then I think that dask.core will remain an easy place to get started for newish developers.